ragna/assistants/_ollama.py

```diff
@@ -0,0 +1,68 @@
+import json
+from typing import AsyncIterator, cast
+
+from ragna.core import RagnaException, Source
+
+from ._api import UnauthenticatedApiAssistant
+
+
+class OllamaApiAssistant(UnauthenticatedApiAssistant):
+    _MODEL: str
+
+    @classmethod
+    def display_name(cls) -> str:
+        return f"Ollama/{cls._MODEL}"
+
+    def _make_system_content(self, sources: list[Source]) -> str:
+        instruction = (
+            "You are an helpful assistants that answers user questions given the context below. "
+            "If you don't know the answer, just say so. Don't try to make up an answer. "
+            "Only use the following sources to generate the answer."
+        )
+        return instruction + "\n\n".join(source.content for source in sources)
+
+    async def _call_api(
+        self,
+        prompt: str,
+        sources: list[Source],
+        *,
+        max_new_tokens: int,
+        api_url: str = "http://localhost:11434/api/chat",
+    ) -> AsyncIterator[str]:
+        async with self._client.stream(
+            "POST",
+            api_url,
+            headers={
+                "Content-Type": "application/json",
+            },
+            json={
+                "messages": [
+                    {
+                        "role": "system",
+                        "content": self._make_system_content(sources),
+                    },
+                    {
+                        "role": "user",
+                        "content": prompt,
+                    },
+                ],
+                "model": self._MODEL,
+                "stream": True,
+                "temperature": 0.0,
+            },
+        ) as response:
+            await self._assert_api_call_is_success(response)
+
+            async for chunk in response.aiter_lines():
+                # This part modeled after https://github.com/ollama/ollama/blob/06a1508bfe456e82ba053ea554264e140c5057b5/examples/python-loganalysis/readme.md?plain=1#L57-L62
+                if chunk:
+                    json_data = json.loads(chunk)
+
+                    if not json_data["done"]:
+                        yield cast(str, json_data["message"]["content"])
+                else:
+                    raise RagnaException("The response was empty.")
+
+
+class Gemma2B(OllamaApiAssistant):
```

Comment:
Not sure about naming here. `vLLM` also supports Gemma. Meaning, without any namespacing, these two cannot coexist under `ragna.assistants`. Maybe we should starting to namespace the builtin assistants with the organization, e.g. `OllamaGemma`, `VllmGemma`, `OpenaiGpt4`, etc?

Let's open an issue for this. I'm ok with merging with the current name if we address the above before next release.

Reply:
I feel better changing it to `OllamaGemma2B` from `Gemma2B` and opening an issue.

Reply:
Discussed in Issue #384